Analysis/WWCategorisation/src/WWCategorisationProcessor.cc

```diff
@@ -0,0 +1,436 @@
+  else if(n_leptons == 1){ // semileptonic
+    if (n_electrons == 1) reco_cat = 2;
+    if (n_muons == 1) reco_cat = 3;
+    if (n_taus == 1) reco_cat = 4;
+  }
+  else if(n_leptons == 0){
+    reco_cat = 1;  // invisible
+    if (_mInv >= 140) reco_cat = 0;  // hadronic
+  }
+
+  _reco_cat_basic = reco_cat;
+
+  // advanced reco categorisation
+  if(reco_cat > 1 && (_nPFO <= 20 || _mInv <= 40)) reco_cat = 5; // any non-hadronic to leptonic
+  if(_nPFO >= 100 || _mInv >= 140) reco_cat = 0; // any to hadronic
+
+  if(reco_cat == 5 && (40 <= _mInv && _mInv <= 140)){  // leptonic to semileptonic
+    if (momtot_max[0] > momtot_max[1] && momtot_max[0] > momtot_max[2]) reco_cat = 2; // sl electron
+    if (momtot_max[1] > momtot_max[0] && momtot_max[1] > momtot_max[2]) reco_cat = 3; // sl muon
+    if (momtot_max[2] > momtot_max[0] && momtot_max[2] > momtot_max[1]) reco_cat = 4; // sl tau
+  }
+
+  if(reco_cat == 0){ // hadronic to ...
+    if(_nPFO <= 20) reco_cat = 5; // leptonic
+    if((20 < _nPFO && _nPFO <= 40) || _misspT >= 40) reco_cat = 1; // semileptonic invisible
+  }
+
+  _reco_cat_advanced = reco_cat;
+
+
+  // store in event parameters
+  pars.setValue("WWCategorisation.TrueCat", _true_cat);
+  pars.setValue("WWCategorisation.RecoCatBasic", _reco_cat_basic);
+  pars.setValue("WWCategorisation.RecoCatAdvanced", _reco_cat_advanced);
+  pars.setValue("WWCategorisation.mInv", _mInv);
+  pars.setValue("WWCategorisation.nPFO", _nPFO);
+  pars.setValue("WWCategorisation.misspT", _misspT);
+  pars.setValue("WWCategorisation.missE", _missE);
+
+  // store in TTree and/or confusion matrix
+  if (_doTT) _observablesTree->Fill();
+  if (_doCM) _confusion_matrix->Fill(true_cat, reco_cat);
+
+  _nTrue[true_cat]++;
+  _nReco[reco_cat]++;
+
+  _nEvt ++;
+}
+
+void WWCategorisationProcessor::check(LCEvent *) {}
+
+void WWCategorisationProcessor::end(){
+  // efficiency confusion matrix
+  for(unsigned int reco_idx = 1; reco_idx < _n_cat+1; reco_idx++){
+    for(unsigned int true_idx = 1; true_idx < _n_cat+1; true_idx++){
+      double val = _nTrue[true_idx-1] != 0 ? _confusion_matrix->GetBinContent(true_idx,reco_idx)/_nTrue[true_idx-1] : 0;
+      _confusion_matrix_ef->SetBinContent(true_idx, reco_idx, val);
+    }
+  }
+
+  if (_doTT){
+    _TTreeFile->cd();
+    _observablesTree->Write();
+  }
+
+  if (_doCM) {
+    TCanvas* can = new TCanvas;
+    PlotConfusionMatrix(can, _confusion_matrix_ef);
+    if (_doTT){
+      _confusion_matrix->Write();
+      _confusion_matrix_ef->Write();
+      can->Write();
+    }
+  }
+
+  if (_doTT) _TTreeFile->Close();
+}
+
+int WWCategorisationProcessor::CountingLeptons(std::vector<int>& daughters_pdg_vector){
+//  int numberof_leptons = 0;
+//  for(long unsigned int i = 0; i < daughters_pdg_vector.size(); i++){
+//    if(( daughters_pdg_vector[i] == 11) || (daughters_pdg_vector[i] == 13) || (daughters_pdg_vector[i] == 15)){
+//      numberof_leptons++;
+//    }
+//  }
+//  return numberof_leptons;
```

Comment:
```suggestion
```
Remove commented code. Otherwise it will become confusing and people have to check whether there is a difference in the commented and running code.

Reply:
done